processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,385 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.groupby.epinephelinae;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Supplier;
+import io.druid.java.util.common.IAE;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.logger.Logger;
+import io.druid.java.util.common.parsers.CloseableIterator;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.query.aggregation.BufferAggregator;
+import io.druid.segment.ColumnSelectorFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.NoSuchElementException;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * A streaming grouper which can aggregate sorted inputs.  This grouper can aggregate while its iterator is being
+ * consumed.  Also, the aggregation thread and iterating thread can be different.
+ *
+ * This grouper is backed by a circular array off-heap buffer.  Reading iterator is able to read from an array slot only
+ * if the write for that slot is finished.
+ */
+public class StreamingMergeSortedGrouper<KeyType> implements Grouper<KeyType>
+{
+  private static final Logger LOG = new Logger(StreamingMergeSortedGrouper.class);
+  // Timeout for waiting for a slot to be available for read/write. This is required to prevent for the processing
+  // thread from being blocked if its iterator is not consumed due to some failure.
+  private static final long DEFAULT_TIMEOUT_MS = 5000L;
+
+  private final Supplier<ByteBuffer> bufferSupplier;
+  private final KeySerde<KeyType> keySerde;
+  private final BufferAggregator[] aggregators;
+  private final int[] aggregatorOffsets;
+  private final int keySize;
+  private final int recordSize; // size of (key + all aggregates)
+  private final long timeoutMs;
+
+  // Below variables are initialized when init() is called
+  private ByteBuffer buffer;
+  private int maxSlotNum;
+  private boolean initialized;
+
+  // Below variables can be read/written by differernt threads
```

Comment:
`volatile` just provides happens-before edges, but doesn't say anything about visibility of the values of the variables themselves. Could you explain more what concurrent pattern of usage of objects of this class is expected?

Reply:
What do you mean by "doesn't say anything about visibility"? From http://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.5, a write to a volatile field happens-before every subsequent read of that field.

The threads reading volatile variables just check their values and can optionally wait for them in a spin loop to become a value if needed.

Reply:
Java Memory Model doesn't have a notion of time. Happens-before relationship describes what will happen *if* the reader thread will read something that is written by writer thread, *but it doesn't guarantee that this read will happen.*

Okay, this is more theoretical than practical, but this concurrent code doesn't seem to be working to me, see below. I suggest to replace it with something probably less efficient but far more predictable (like a queue), and optimize later